File: docs/research/corpus-examples/auth-session/php/vulnerable-2.php
<?php
// Source: InvoicePlane/InvoicePlane, application/modules/sessions/controllers/Sessions.php,
// commit a479033943de0de61b62743fd6abb6d277b69f2d (the commit immediately BEFORE the
// aeddda43... fix below, in the same file's history). License: MIT (code); the
// "InvoicePlane" name/logo carry a separate trademark restriction that does not apply to
// this source excerpt (see the repo's LICENSE.txt).
//
// Excerpt from Sessions::forgotPassword(), CodeIgniter 3 controller. Trimmed to the
// token-generation step and its surrounding control flow; the full method also does bot
// detection and IP/email rate limiting (omitted here as unrelated to the token weakness).
//
// Vulnerability (CWE-330 predictable value / CVE-2021-29023): the reset token is
// md5(time() . $email . $this->crypt->salt()). $email is known to the attacker (it's the
// address they're attacking) and time() is a low-entropy, narrow-window value guessable to
// within seconds from the HTTP response's own Date header -- so the only real secret input
// is $this->crypt->salt(), and the *comment* claims that salt exists specifically "to
// prevent predictability" (CVE-2021-29023). MD5 of a mostly-known/guessable input, rather
// than a token drawn from a CSPRNG, remains brute-forceable within the small remaining
// keyspace once email+time are pinned down.

class Sessions extends Base_Controller
{
    public function forgotPassword()
    {
        // ... email validation, bot detection, and rate-limit checks omitted ...

        $email = $this->input->post('email', true);

        // Test if a user with this email exists
        $this->db->where('user_email', $email);
        $user = $this->db->get('ip_users')->row();

        // Security: Always show the same message regardless of whether email exists
        // This prevents email enumeration attacks
        if ($user) {
            // User exists - send actual reset email
            //use salt to prevent predictability of the reset token (CVE-2021-29023)
            $this->load->library('crypt');
            $token = md5(time() . $email . $this->crypt->salt());

            // Save the token to the database
            $db_array = [
                'user_passwordreset_token' => $token,
            ];

            $this->db->where('user_email', $email);
            $this->db->update('ip_users', $db_array);

            // Send the email with reset link
            $this->load->helper('mailer');
            $email_resetlink = site_url('sessions/passwordreset/' . $token);
            $email_message   = $this->load->view('emails/passwordreset', [
                'resetlink' => $email_resetlink,
            ], true);

            // ... mail-sending branch omitted ...
        }

        // ... generic "check your email" response omitted ...
    }
}

File: docs/research/corpus-examples/auth-session/python/vulnerable-2.py
# Source: DOAJ/doaj, portality/view/account.py, commit db2511f61e9033d10a3ea6d9fc90fa122b9e068a.
# License: Apache-2.0.
#
# Excerpt: handle_login_code_request() only, trimmed from the full account.py view module
# (Flask-Login / Flask-WTF imports, other account routes, and the DOAJ event/messaging
# service calls used for the actual email send are omitted as unrelated to token
# generation).
#
# Vulnerability (CWE-330 use of insufficiently random values / CWE-338 weak PRNG): the
# passwordless-login code sent to the user's email is 6 digits built from
# random.randint(0, 9) -- Python's `random` module is a Mersenne Twister, not a CSPRNG, and
# is not intended for security tokens (its state can be reconstructed from a sequence of
# outputs). Combined with only 10**6 = 1,000,000 possible codes and no visible rate-limit
# in this excerpt, the login code is both statistically weak *and* small enough to be
# brute-forced online within the 10-minute LOGIN_CODE_TIMEOUT if no separate throttling
# exists on the verification endpoint. Contrast with the same file's own password-reset
# token (idiomatic-2.py), which uses uuid.uuid4().hex -- a CSPRNG-backed 128-bit value.
import random


def handle_login_code_request(user, form):
    LOGIN_CODE_LENGTH = 6
    LOGIN_CODE_TIMEOUT = 600  # 10 minutes

    code = ''.join(str(random.randint(0, 9)) for _ in range(LOGIN_CODE_LENGTH))
    user.set_login_code(code, timeout=LOGIN_CODE_TIMEOUT)
    user.save()

    # ... svc.send_login_code_email(user, code, ...) and template render omitted ...
    return code
